### PR TITLE
Additional -Maker properties on Config for GAE support

### PR DESCRIPTION
**Pull request by buu700**
I tried to keep everything about these new properties as consistent with CookieStoreMaker/SessionStoreMaker as I could. However, I may very well have structured this in such a way that the new -Maker functions won't always have been called by the time they're needed to have been called. I didn't do any testing or thorough investigation beyond some greps to make sure that my change wasn't obviously wrong, as I figure that with your better grasp on the broader structure and context of the code, issues of that nature will be more immediately apparent to you when you review this.

### CODE REVIEW

**Comment by aarondl** [comment]
Hey @buu700, how's this PR going? I was going to look at it this weekend. Is it ready for review?

**Comment by buu700** [comment]
Hey Aaron, yep, was just leaving a comment. These are the last few changes that I needed for our project to build and run without errors, which it now does in my testing. I tried to keep the changes conservative, but there were a bunch of minor issues that I had to address, so totally understood if this isn't the direction you'd want to go in and/or some of what I did was unnecessary.

**Comment by buu700** [comment]
Also, since I should probably explain my changes specifically:

* ModuleName properties: GAE rejects code that doesn't pass `go vet`, which includes "unused imports" irrespective of whatever side effects those imports may be producing, so the cleanest workaround seemed to be to give the code importing Authboss the option to pass X.ModuleName into Init to get rid of the warning.

* Making NewContext do a deep clone and having all the modules reference the properties of current context instead of their own properties: I made this change when I suspected a run-time error in my code to be caused by the previous style causing a concurrency issue, but it turned out that it was just a goroutine causing that error. That said, intuitively, it seems like this would be necessary anyway, and it's consistent with the fact that SessionStorer and CookieStorer are always referenced from the context. Actually, now that I think about it, the deep clone part seems unnecessary for the same reason it wasn't necessary with Session/CookieStoreMaker; I'll revert that part unless there's some reason you want to keep it.

* DisableGoroutines option: sending mail from a goroutine was upsetting App Engine.

File: config.go [hunk 1]
```diff
@@ -21,6 +21,8 @@ type Config struct {
 	RootURL string
 	// BCryptCost is the cost of the bcrypt password hashing function.
 	BCryptCost int
+	// If true, authboss won't use any goroutines. Dependencies of authboss may or may not use goroutines.
+	DisableGoroutines bool
```
Comment (aarondl) [inline config.go]: I think the comment should read:
"DisableGoroutines stops authboss from spawning any goroutines inside the library for example to send mail".

Also, I'd like to push back on this flag, it's an anti-feature. Enabling it makes your app strictly worse as it doesn't follow best practices of isolating your background tasks from your request response. Is GAE really this limiting? Do we have no other workaround?

File: confirm/confirm.go [hunk 2]
```diff
@@ -130,17 +132,21 @@ func (c *Confirm) afterRegister(ctx *authboss.Context) error {
...
+	if ctx.DisableGoroutines {
+		c.confirmEmail(ctx, to, token)
+	} else {
+		go c.confirmEmail(ctx, to, token)
+	}
 }
 
 // confirmEmail sends a confirmation e-mail.
-func (c *Confirm) confirmEmail(to, token string) {
+func (c *Confirm) confirmEmail(ctx *authboss.Context, to, token string) {
```
Comment (aarondl) [inline confirm/confirm.go]: This addition causes the tests to fail as it's changed the API. Please run tests on all pull requests. Ideally we'd have CI for this but it never got set up.

File: recover/recover.go [hunk 3]
```diff
@@ -194,11 +196,15 @@ func newToken() (encodedToken, encodedChecksum string, err error) {
...
+var goRecoverEmail = func(r *Recover, ctx *authboss.Context, to, encodedToken string) {
+	if ctx.DisableGoroutines {
+		r.sendRecoverEmail(ctx, to, encodedToken)
+	} else {
+		go r.sendRecoverEmail(ctx, to, encodedToken)
+	}
 }
 
-func (r *Recover) sendRecoverEmail(to, encodedToken string) {
+func (r *Recover) sendRecoverEmail(ctx *authboss.Context, to, encodedToken string) {
```
Comment (aarondl) [inline recover/recover.go]: This is the change that breaks the tests here.

```text
recover/recover_test.go:271: not enough arguments in call to r.sendRecoverEmail
recover/recover_test.go:288: not enough arguments in call to r.sendRecoverEmail
```

File: confirm/confirm.go (hunk 2 above)
Comment (aarondl) [inline confirm/confirm.go]: Sorry, my bad this isn't the breakage. It just looked similar. Annotated the real failure below.

**Comment by aarondl** [comment]
Happy to accept this with a few sticking points.

The DisableGoroutines as I mentioned is an anti-feature. It'd be great if we could come up with something better. If not, it'll have to do.

The `ModuleName` change doesn't sit right with me. I don't see any go vet errors? Can you be more specific what problems you're encountering. I pushed a change that fixes the go vet errors throughout the project (there were some, but they're resolved now).

**Comment by buu700** [comment]
Thanks, and sorry about that! I'd thought that I'd gotten in all the necessary updates to the tests, but I'll confirm that they actually build and pass this time.

As far as DisableGoroutines, I don't like it either, but what seems to happen is that the current request and context objects are expired by the time any App Engine APIs are called, causing the API calls to fail. As an alternative, how about I get rid of DisableGoroutines and replace the logic that currently references it with checks for the existence of the relevant -Maker property (since it can be assumed that any code that opts for the -Maker version has a hard dependency on a valid request/context)?

As far as ModuleName, to be specific, it isn't Auhtboss itself that fails go vet, but rather my code that depends on Authboss. While each module import isn't truly unused in the sense that it produces side effects that my code depends on, go vet doesn't take side effects into account. I wasn't sure what the best way to handle that was, but it seemed cleaner than any alternative I could think of to have each module expose that constant for me to explicitly pass it into Authboss.Init.

**Comment by aarondl** [comment]
Yeah. If we document the presence of the relevant -Makers and people see it
as a side effect of enabling them it's probably cleaner.

Also I'm still confused about the module name part because if you look at
authboss-sample it does these same imports yet has no vet errors. Without
seeing the errors and some code I'm still not sure about that part.

Getting closer to merge! Sorry to make all this extra work for you. lol
On May 9, 2016 9:11 AM, "Ryan Lester" <notifications@github.com> wrote:

Thanks, and sorry about that! I'd thought that I'd gotten in all the
necessary updates to the tests, but I'll confirm that they actually build
and pass this time.

As far as DisableGoroutines, I don't like it either, but what seems to
happen is that the current request and context objects are expired by the
time any App Engine APIs are called, causing the API calls to fail. As an
alternative, how about I get rid of DisableGoroutines and replace the logic
that currently references it with checks for the existence of the relevant
-Maker property (since it can be assumed that any code that opts for the
-Maker version has a hard dependency on a valid request/context)?

As far as ModuleName, to be specific, it isn't Auhtboss itself that fails
go vet, but rather my code that depends on Authboss. While each module
import isn't truly unused in the sense that it produces side effects that
my code depends on, go vet doesn't take side effects into account. I wasn't
sure what the best way to handle that was, but it seemed cleaner than any
alternative I could think of to have each module expose that constant for
me to explicitly pass it into Authboss.Init.

—
You are receiving this because you commented.
Reply to this email directly or view it on GitHub
<https://github.com/go-authboss/authboss/pull/83#issuecomment-217910748>

**Comment by buu700** [comment]
Cool, I'll get rid of DisableGoroutines and add some extra info to the -Maker documentation.

As far as the go vet issue, thanks! The solution from authboss-sample is working for me; it never occurred to me that one could just name an import `_` to explicitly mark it as not intended for direct use in the current file. I'll revert that ModuleName thing as well.

Also, how do I run the tests? `go test recover/recover_test.go` is just complaining about not being allowed to reference the `internal` module.

**Comment by aarondl** [comment]
You'll have to make sure that you've cloned into `$GOPATH/src/
gopkg.in/authboss.v0` and then `go test ./...`

Also it seems like maybe you're not using goimports as a replacement for
gofmt? Else it would delete any unused imports and the only way to get them
to stay is to name the import like this. Very nice little program as it
auto adds and deletes your imports for you. Just thought I'd mention it in
passing because I can't imagine manually handling imports anymore. Might
want to add it to your dev environment! :)

**Comment by buu700** [comment]
Got it, thanks for the help! Everything should be good to merge now unless something new stands out as needing fixing.

And thanks for the goimports recommendation! I'll go ahead and swap out gofmt for that in our dev environment.

**Comment by buu700** [comment]
(Just shortening the comment so GitHub doesn't cut it off.)

**Comment by buu700** [comment]
Hm, so completely unrelated to this PR, but oddly enough goimports removed a bunch of stuff that was actually being used and just caused our code to stop compiling. Guess I'll stick with gofmt for now until I have a chance to figure that out.

**Comment by aarondl** [comment]
GOPATH is important for goimports as well. Make sure it's set properly in your terminal/editor when you use it. Thanks for the PR.

**Comment by buu700** [comment]
Ah, cool, I'll look into that. Thanks for the merge!

**Pull request merged**